### DISCUSSION TITLE
Setting the quarkus.container-image.labels ?

### DISCUSSION THREAD

**Discussion by nielsbasjes** [quarkusio/quarkus#26633] [Q&A] [answered]
I want to tag my native image conforming to these labels https://github.com/opencontainers/image-spec/blob/main/annotations.md

The `quarkus.container-image.labels` are defined as a `Map<String,String>`.
https://quarkus.io/guides/container-image#quarkus-container-image_quarkus.container-image.labels-labels

Note: I expected this page https://quarkus.io/guides/config-reference to have info on setting a `list of strings` and a `Map<String, String>` in the various available places. Also the ways of setting values in the pom.xml seems completely missing. After much searching I found an example for a different `Map` which simple appends the `key` of the map entry after the property name.

So in my pom.xml I put these properties (among many others)
```
<quarkus.container-image.labels.vendor>Niels Basjes</quarkus.container-image.labels.vendor>
<quarkus.container-image.labels.org.opencontainers.image.vendor>Niels Basjes</quarkus.container-image.labels.org.opencontainers.image.vendor>
```

Now when I build my native image and do a `docker inspect` of that image I see (summary):

```
"Labels": {
    "architecture": "x86_64",
...
    "com.redhat.build-host": "cpt-1008.osbs.prod.upshift.rdu2.redhat.com",
    "com.redhat.component": "ubi8-micro-container",
...
    "vendor": "Niels Basjes",
}
```

Simply put: 
- A docker image label name allows the '.' in the label name as is clearly visible from the inherited labels
- The labels I have defined with a '.' in the name are not included, the ones without a '.' ARE included.

What is the proper way to configure this?

---------

As an extra test (to try and see what works) I have added these to my `./config/application.properties` 

```
quarkus.container-image.labels["name1"]                                   = "yauaa1"
quarkus.container-image.labels.name2                                   = "yauaa2"
quarkus.container-image.labels["org.opencontainers.image.url1"]           = "https://yauaa.basjes.nl/1"
quarkus.container-image.labels.org.opencontainers.image.url2           = "https://yauaa.basjes.nl/2"
```

Only the quarkus.container-image.labels.name2 was present in the image.

**Comment by albers** [comment, marked as answer]
Configuring image labels works similar to [configuring log levels](https://quarkus.io/guides/logging#logging-categories).
If a map key (logging category or image label) contains dots, you have to quote the key.

Thus, in order to create an image label `org.opencontainers.image.url` with the value _https://some.host/_, you have to use

```config
quarkus.container-image.labels."org.opencontainers.image.url"=https://some.host/
```

**Reply by nielsbasjes** [reply]
Thanks.
Does this (having a " in the key) mean that these cannot be set as properties in the pom.xml?

**Reply by nielsbasjes** [reply]
Thanks this works.

For others trying to do the same:

In my project I have moved `src/main/resources/application.properties` to `src/main/config/application.properties` and it now has lines like these:
```
quarkus.container-image.labels."org.opencontainers.image.authors"       = Niels Basjes
quarkus.container-image.labels."org.opencontainers.image.created"       = ${maven.build.timestamp}
quarkus.container-image.labels."org.opencontainers.image.url"           = https://yauaa.basjes.nl
quarkus.container-image.labels."org.opencontainers.image.version"       = ${project.version}
```

My pom.xml has these two additions:
```
<resources>
    <resource>
      <directory>src/main/config</directory>
      <filtering>true</filtering>
    </resource>
  <resource>
    <directory>src/main/resources</directory>
  </resource>
</resources>
```

```
<!-- Explicitly needed because of https://stackoverflow.com/a/52939311/114196 -->
<plugin>
  <groupId>org.apache.maven.plugins</groupId>
  <artifactId>maven-resources-plugin</artifactId>
  <version>3.2.0</version>
</plugin>
```

When I run the build and do `docker inspect` I see the expected labels:

```
"org.opencontainers.image.authors": "Niels Basjes",
"org.opencontainers.image.created": "2022-07-12T07:57:55Z",
"org.opencontainers.image.url": "https://yauaa.basjes.nl",
"org.opencontainers.image.version": "7.3.1-SNAPSHOT",
```

**Comment by albers** [comment]
> Does this (having a " in the key) mean that these cannot be set as properties in the pom.xml?

I think so. A property name in a POM file is syntactically an XML element, and according to https://www.w3.org/TR/xml/#NT-NameChar, `"` is not allowed in element names.

We do have a more general problem here. Quarkus configuration relies on mapping a key/value store on various configuration sources.
Some configuration sources have limitations on key names, so special encoding rules exist (e.g. replacing dots and dashes in environment variables with underscores).
Using map configuration keys adds another dimension of problems. You basically serialize arbitrary key names into the key names of the Quarkus configuration, which causes collisions with the naming conventions of the configuration sources. So additional encoding tricks are required.

Here is what I found out so far

- In property files, quoting the embedded map key works: `quarkus.container-image.labels."org.opencontainers.image.version"=1.0`
- For system properties, escaped quotes work: `-Dquarkus.container-image.labels.\"org.opencontainers.image.version\"=1.0`

I found no solution for maven properties and environment variables. According to https://smallrye.io/smallrye-config/2.10.1/config/environment-variables/ , SmallRye Config provides additional conversion rules for maps, but I did not manage to set any container label via env vars, even with non-colliding names like `QUARKUS_CONTAINER_IMAGE_LABELS_FOO=baz`.
